WIP: Overview Documentation for PSF Fitting Workflow

Onoddil opened the pull request:
This pull request is set up for the discussion of the overall PSF fitting process documentation, the fitting workflow and how each block interacts internally, and should not be merged until after the discussion is finished. This documentation complements that in #721 where the documentation for those blocks implemented, with solid APIs unlikely to change in future releases, and the newer APIs available in PRs #753, #755 and #756 for which details may be subject to change.

The main issues for discussion in this document are mostly limited to formatting and clarity of wording, as discussion of individual blocks should be directed to their respective PRs. The issues are summarised below:

- [ ] High Level Overview
  - [ ] Name
  - [ ] Wording
  - [ ] Formatting
  - [ ] Workflow graphic representation

One issue that should be discussed is the final graphical representation of the workflow. The current graph is simply a reproduction of the [block diagram](https://github.com/astropy/photutils/blob/master/docs/psf_spec/block_diagram.png), but this PR is meant to supersede that (and closes #761), and so confirmation of the workflow represented here is merited.

Please provide any feedback on the description of the overall PSF fitting process you may have, such that the implementation meets all of the requirements of all users and is as clear as possible going forward. Attached is the final output of the [graph](https://user-images.githubusercontent.com/3527254/48089546-c3410480-e1d2-11e8-93ed-96e8ede31368.jpeg) for viewing, as currently ``graphviz`` is not viewable within this PR.

eteq commented on docs/psf_spec/high_level_overview.rst:
```diff
@@ -0,0 +1,244 @@
+        gs [label = "guess at stars", shape = box];
+        p [label = "psf model"];
+        sm [label = "scene maker"];
+        st [label = "stars\n(astropy.table)", shape = box];
```
I wonder of "stars" here should be "objects" or something like that?  Presumably the scene maker at least *might* have the option to do other kinds of fitting?

marthaboyer replied:
I second this also

eteq commented on docs/psf_spec/high_level_overview.rst:
```diff
@@ -0,0 +1,244 @@
+PSF fitting framework, providing a top-down view of the process. Where available, links
+to more detailed descriptions of each block is also provided.
+
+.. graphviz::
```
In general this is great!  One thing about this, though: It might be helpful to have some lines that pass through the image.  That is, have lines from where the image "enters" the big box to the finder and the background estimator?

And another thing: maybe try to add a "legend" that shows what the lines mean?  E.g. "solid = required flow of data", "dashed = optional information flow" and "dotted = optional component" or something like that?

marthaboyer replied:
I'll second this. I was also going to suggest a description of what the different lines mean.

eteq commented on docs/psf_spec/high_level_overview.rst:
```diff
@@ -0,0 +1,244 @@
+Input Image
+^^^^^^^^^^^
+
+The input image is the most fundamental part of the PSF fitting process, as it is the
```
Should probably say something here about `NDData` objects.  That is, the fundamental object we will want to use to pass around the images is probably `NDData`.  There can just be something then that says roughly "but there's a way in that could instead use raw arrays or FITS files that then get auto-converted to `NDData"?  Or that might be a detail that can instead be in the other documents.

marthaboyer replied:
Since there are details on the content of the image here, I agree with Erik that NDData objects should be mentioned.

eteq commented on docs/psf_spec/high_level_overview.rst:
```diff
@@ -0,0 +1,244 @@
+Finder
+^^^^^^
+
+The finder is the first step in the PSF fitting process, as sources must be discovered in
```
These all need to be implementation of the `StarFinderBase` class, right? (i.e., https://photutils.readthedocs.io/en/stable/api/photutils.StarFinderBase.html#photutils.StarFinderBase).  If so you should say that, and link to it (e.g. ``  `~photutils.detectionStarFinderBase`  ``)

eteq replied:
More broadly: I wonder if the `StarFinderBase` should actually be `ObjectFinderBase` or something like that?  Not necessary for this document, but worth thinking a bit about.

Onoddil replied:
I've updated the above references of ``stars`` to ``sources`` as that doesn't particularly affect anything and can be future-proofed now, but I obviously can't change these names as they're hard-coded links for now. Where would be best to raise this as a separate issue? Is actually making an issue the right approach for this?

eteq commented on docs/psf_spec/high_level_overview.rst:
```diff
@@ -0,0 +1,244 @@
+^^^^^^^^^^^
+
+The second block to run in the fitting process, the various group maker processes, such as
+`~photutils.psf.DAOGroup`, allow for the merging of astrometrically overlapping sources
```
```suggestion
 `~photutils.psf.DAOGroup`, allow for the merging of sources that are astrometrically nearby
```

"overlapping" is a bit unclear because the definition of "overlapping" somewhat depends on the algorithm.  I think "nearby" is better in that it's intentionally vague?

eteq commented on docs/psf_spec/high_level_overview.rst:
```diff
@@ -0,0 +1,244 @@
+The second block to run in the fitting process, the various group maker processes, such as
+`~photutils.psf.DAOGroup`, allow for the merging of astrometrically overlapping sources
+into a set of multiple stars. These stars must be fit simultaneously (see Fitter_) as a
+composite model, avoiding the potential issue where blended sources return incorrect fluxes,
```
```suggestion
 composite model, making deblending of sources possible.
```
(contingent on the next line suggestion)

eteq commented on docs/psf_spec/high_level_overview.rst:
```diff
@@ -0,0 +1,244 @@
+`~photutils.psf.DAOGroup`, allow for the merging of astrometrically overlapping sources
+into a set of multiple stars. These stars must be fit simultaneously (see Fitter_) as a
+composite model, avoiding the potential issue where blended sources return incorrect fluxes,
+as affects aperture photometry. The block accepts the `~astropy.table.Table` output from
```
```suggestion
The block accepts the `~astropy.table.Table` output from
```

eteq commented on docs/psf_spec/high_level_overview.rst:
```diff
@@ -0,0 +1,244 @@
+Fitter
+^^^^^^
+Once a set of sources has been detected and grouped, each source must have its respective
+properties -- position and flux primarily -- determined. For this a ``Fitter`` instance is
```
```suggestion
 properties -- always position and flux, but sometimes other properties -- determined. For this a ``Fitter`` instance is
```

eteq commented on docs/psf_spec/high_level_overview.rst:
```diff
@@ -0,0 +1,244 @@
+Once a set of sources has been detected and grouped, each source must have its respective
+properties -- position and flux primarily -- determined. For this a ``Fitter`` instance is
+required; these are drawn from a separate class or instance of minimization routines, such as
+those given under the `~astropy.modeling` umbrella. The fitter must be a two-dimensional fitter,
```
```suggestion
 those implemented in `~astropy.modeling.fitting`. The fitter must be a two-dimensional fitter,
```

eteq commented on docs/psf_spec/high_level_overview.rst:
```diff
@@ -0,0 +1,244 @@
+required; these are drawn from a separate class or instance of minimization routines, such as
+those given under the `~astropy.modeling` umbrella. The fitter must be a two-dimensional fitter,
+by nature of the image to be fit, accepted as an input to the instance. The class must then
+return the position of the source (or sources, if a composite model was supplied; see
```
I think this part on the exact inputs/outputs could be left out because it's actually a bit more complicated (i.e., it has to conform to the `astropy.modeling.fitting` fitter API).  So just saying "it has to take in the image and table and output the fitted model.  More detail given in ..." or something like that is sufficient.

eteq commented on docs/psf_spec/high_level_overview.rst:
```diff
@@ -0,0 +1,244 @@
+
+PSF Model
+^^^^^^^^^
+Perhaps the most important aspect of PSF fitting, the ``PSF Model`` describes the distribution
```
```suggestion
 Perhaps the most variable aspect of PSF fitting, the ``PSF Model`` describes the distribution
```
?

Onoddil replied:
I was trying to lead into this section with the idea that while the rest of the sections (getting good sources, grouping blended things, etc.) are important, *this* section has to be done right or there's no point doing any of the other stuff at all... So important is what I meant (instead of variable, as in "most likely to change for each user"?), but this is slightly qualitative and perhaps subjective language; none of the other sections have lead-in.

I think the question is more:
do we just start this section with ```The ``PSF model`` describes the distribution```?

eteq commented on docs/psf_spec/high_level_overview.rst:
```diff
@@ -0,0 +1,244 @@
+Perhaps the most important aspect of PSF fitting, the ``PSF Model`` describes the distribution
+of light falling on a given pixel from a source at some position with some total flux, fully
+describing the effects of telescope optics, quantized CCD pixels, etc. In this respect it is
+more properly described as a PRF than a PSF -- see the :doc:`documentation <psf_model>` for
```
I think it would be good to also link to the "terminology" section here: https://photutils.readthedocs.io/en/latest/psf.html#terminology (that is, link using sphinx syntax to the point in the docs that that link goes to), since that discusses the PRF/PSF/etc thing.

marthaboyer replied:
agreed. Or at the very least, define the acronyms.

eteq commented on docs/psf_spec/high_level_overview.rst:
```diff
@@ -0,0 +1,244 @@
+of light falling on a given pixel from a source at some position with some total flux, fully
+describing the effects of telescope optics, quantized CCD pixels, etc. In this respect it is
+more properly described as a PRF than a PSF -- see the :doc:`documentation <psf_model>` for
+more details -- but we conflate the two terms here freely. Similar to the ``Fitter``, it is
```
```suggestion
 more details -- but we use a single term here for brevity. Similar to the ``Fitter``, it is
```
we should probably fix this at some point and use the terminology more consistently?  But that can be done later, probably.

Onoddil replied:
I've made the change to the language, but to the more overall point: I'm not sure we really *do* need to change the terminology. Literally nowhere in the `photutils` framework do we NOT do PRF fitting, by nature of fitting quantised images we can't do anything but. Thus so long as we make it clear that technically PSFs can be (and are) pedantically different and that we are actually doing PRF fitting but we're just gonna use the colloquialism if that's ok with everyone else I think we get away with it.

Instead of putting loads of work into rewriting all of our documentation to use the new and more proper language, we just have to put work into making sure at any necessary point we differentiate between the two and link to the terminology page (like above!).

eteq commented on docs/psf_spec/high_level_overview.rst:
```diff
@@ -0,0 +1,244 @@
+PSF, such as that implemented by the effective PSF functionality (see :ref:`build-epsf` for
+details). The model must describe a two-dimensional distribution, and must return the total
+flux of the source and its sub-pixel position via the ``evaluate`` function, accepted by the
+Fitter_ when minimizing the fit between the model and the data.
```
As above with the fitter, I think it's best to describe it as it has to follow the `astropy.modeling.models` API for 2D models, because there's a lot of "little things" there that may end up mattering a lot.

eteq commented on docs/psf_spec/high_level_overview.rst:
```diff
@@ -0,0 +1,244 @@
+^^^^^^^^^^^^^^^^^
+The ``noise`` block within the PSF fitting process is, in much the same way as the ``Fitter``
+block, a separate ``callable`` function mapping the relationship between the pixel values in
+the input ``image`` and the corresponding uncertainty -- in units of standard deviations -- of
```
I'm not sure I like the "in units of standard deviation part", but will discuss more in #755

Onoddil replied:
I've removed the subclause for now as it's not particularly necessary (this being a brief, high level overview and all) but we can discuss that further in the other PR and I'll make any changes finalised there in this branch too.

eteq commented on docs/psf_spec/high_level_overview.rst:
```diff
@@ -0,0 +1,244 @@
+The ``image`` on which source extraction and evaluation is to be performed is not necessarily
+assumed to have had any pre-processing applied to it, and thus it may be necessary to account
+for the counts of the image background when handling PSF fitting. This block handles
+the determination of the background levels, avoiding high count peaks to calculate the typical
```
what do you mean by "avoiding high count peaks"?  I think that might be algorithm-specific, so maybe better to just do this:
```suggestion
 the determination of the background levels, and should provide the typical
```
?

eteq commented on docs/psf_spec/high_level_overview.rst:
```diff
@@ -0,0 +1,244 @@
+
+Culler and Ender
+^^^^^^^^^^^^^^^^
+Only implemented in iterative fitting classes, part of the extension to
```
I don't understand the "part of the extension to" part... partly because `BasicPSFPhotometry` isn't really defined here (see my general comment)

eteq commented on docs/psf_spec/high_level_overview.rst:
```diff
@@ -0,0 +1,244 @@
+block therefore requires as its inputs the PSF model and the object type -- a string representing
+the type of source (point-like, spiral galaxy, etc.) to create the intrinsic light distribution
+for -- and returns a new `~astropy.modeling.Fittable2DModel` instance, the convolution of the PSF
+and the true source. Further specifics on this block, including how it will link to `Scene Maker`_
```
```suggestion
 and the true source. Further specifics on this block, including how it connects to `Scene Maker`_
```

eteq commented on docs/psf_spec/high_level_overview.rst:
```diff
@@ -0,0 +1,244 @@
+of both point sources into an overlapping, simultaneous fit group but also into a single extended
+object. Thus each iteration will require a step in which multiple-point-source extended objects are
+evaluated for their separation, and newly detected point sources are evaluated for potential
+assignment as part of an extended source. The details of this block are available :doc:`here
```
This may be better discussed in #756 ... but this discussion makes me realize that maybe the scene maker and single object model should be one thing?  Or even if separate classes, maybe they should be described together in one document?

Onoddil replied:
They're basically "point + extended" sources extensions of group maker and the PSF model, so while linked slightly more obviously by the need to share a common ``object_type`` list and know what to do with each available choice in each instance, they're no more or less inter-connected than group maker and PSF model; it's just obvious what to do with those as the only potential object type is a point source, which is simple to handle.

We can discuss this further in the corresponding PR to get into the details, and, again, I'll make any changes decided upon in there in this branch too.

eteq commented on docs/psf_spec/high_level_overview.rst:
```diff
@@ -0,0 +1,244 @@
+
+.. _Single Object Model:
+
+Single Object Model
```
This (along with the scene maker) are *optional*, right?  That is, if not present, it means just use the PSF model directly?  In that case you should say that in both descriptions.

Onoddil replied:
I've added language to this effect in the text, but eventually I don't think they'll be entirely optional, in that I think the block will always have to be used, but it can have convenient "do nothing" defaults. I think it's tied to the ``object_type`` flag that will eventually have to be thrown around in the ``astropy.table.Table`` objects: if that (by default) is something like ``point source`` then ``Single Object Model`` simply returns the input PSF model (i.e., f * &#948; = f) and ``Scene Maker`` just returns the input table of objects (i.e., no input sources are an extended source split apart).

This *effectively* means they're optional, as the user can just ignore that the more complex functionality exists and default parameters unchanged give the same answer as if they weren't there, but I think that's easier than having ``if use_som: new_model = single_object_model(psf)`` or equivalent.

marthaboyer commented on docs/psf_spec/high_level_overview.rst:
```diff
@@ -0,0 +1,244 @@
+:raw-html:`&rarr;` ``fitter`` :raw-html:`&rarr;` ``culler and ender`` :raw-html:`&rarr;`
+``finder`` :raw-html:`&rarr;` ... :raw-html:`&rarr;` ``stars``.
+
+Each submodule is described briefly here, with context for how it fits into the overall
```
Minor fix to Erik's suggestion: A brief summary of each block is described here...

marthaboyer commented on docs/psf_spec/high_level_overview.rst:
```diff
@@ -0,0 +1,244 @@
+:raw-html:`&rarr;` ``fitter`` :raw-html:`&rarr;` ``culler and ender`` :raw-html:`&rarr;`
+``finder`` :raw-html:`&rarr;` ... :raw-html:`&rarr;` ``stars``.
+
+Each submodule is described briefly here, with context for how it fits into the overall
```
Minor fix to Erik's suggestion: A brief summary of each block is described here...

marthaboyer commented:
@Onoddil This all looks fantastic to me! I agree with Eriks suggestions & don't really have much else to add.

Onoddil commented:
Thanks @eteq and @marthaboyer for proofs! I've made most of these changes without comment as they're good suggestions, but I've replied to a few inline comments for further question.

With regards to the larger comment of how ``BasicPSFPhotometry`` interacts with ``IterativelySubtractedPSFPhotometry``, I'm not sure how this should be tackled. I don't like that we focus so much on a single iterative fitting algorithm (as of writing it's the only one available, but see #732 for a new one...). The concept is that a "Basic PSF fit" is (see the second paragraph of this document for the basic flow, where I thought I discussed this but perhaps not obviously enough): ``image`` -> ``finder`` (optionally) -> ``grouping`` -> ``fitting`` -> ``sources``. An "Iterative PSF fit" does those things, then goes to ``culler and ender``, potentially interacts with ``image`` again (``IterativelySubtractedPSFPhotometry``, by the name, subtracts sources found in one iteration and re-fits the residuals, e.g.), potentially interacts with ``guess at sources`` (#732 stacks all previously detected sources to be re-fit together, instead of fitting sources in residuals, so it swaps image editing for output table editing), then goes back to ``finder``... Thus one could *actually* view a "Basic PSF fit" as an "Iterative PSF fit" with ``culler and ender`` defined to never cull any sources but always end the 'loop'. 

However, as I imagine we can't really get rid of ``BasicPSFPhotometry`` (as much as I'd like to just define a "Basic PSF fit" as "pick your favourite iterative routine but don't worry, set ``Culler and Ender`` to this ``BasicCullerEnder`` and it'll not make any difference"), I think we could better factor the codebase to use as much of ``BasicPSFPhotometry`` in subsequent subclasses as possible to highlight that it's basically just one iteration of the code. Thus while a ``BasicFit`` is ``image`` -> ``finder`` (optionally) -> ``grouping`` -> ``fitting`` -> ``sources``, ``IterativePSFPhotometry`` is ``BasicFit`` -> ``culler and ender`` -> (maybe end or) -> ``mess with image`` -> ``BasicFit`` -> ... -> end -> ``sources``.

As currently written I don't think it's quite so simple but I don't think the refactor to define things as such is too difficult; ``BasicPSFPhotometry`` and iterative subclasses do the heavy lifting in ``do_photometry`` but if that were changed so ``BasicPSFPhotometry`` essentially has ``def do_photometry(self, ...): return do_single_loop(self, ...)`` then subclasses could overload ``do_photometry`` but still use the super ``do_single_loop`` function.

For this documentation (instead of me waxing philosophical about what I'd like to do to the code...): the attempt to explain how ``IterativePSFPhotometry`` builds off ``BasicPSFPhotometry`` is in paragraph two. I added the box around the basic PSF parts of the block diagram, but it's basically everything except the culler and ender. I also added the part about "messing with the image" as an optional block. I think (given that I've just reproduced someone else's original block diagram) that the "subtract" floating between fitter and culler and ender was ``IterativelySubtractedPSFPhotometry``'s "subtract fit sources from image and find more in the residuals", so we should actually probably more generalise that either way, as that isn't true for all algorithms.

Let me know if the improvements to the block diagram help, and then any suggestions for how to better explain that ``IterativePSFPhotometry`` is the entirety of ``BasicPSFPhotometry`` + ``Culler and Ender`` + ``Tweak Input Image`` as a single iteration.

Onoddil commented:
Hi @marthaboyer and @eteq - as per a conversation I had offline with @eteq I've pushed the changes I made in response to the comments that were straightforward, but have a few inline replies to comments for clarifications, and the more general query in the comment above. I would appreciate it if you could review the changes I already made, have a look at the queries I had to a few of the comments you had previously, and have a think about the larger issue in my reply above this one.

larrybradley commented:
The 2017-2018 PSF photometry API plans (including the additions in this PR) have been archived on the `photutils` wiki page:  https://github.com/astropy/photutils/wiki/PSF-photometry-plans-(2017%E2%80%902018)
